mbs_results/ratio_of_means.py

```diff
@@ -0,0 +1,388 @@
+    auxiliary: str,
+    filters: pd.DataFrame = None,
+) -> pd.DataFrame:
+    """
+    Imputes for each non-responding contributor a single numeric target
+    variable within the dataset for multiple periods simultaneously. It uses
+    the relationship between the target variable of interest and a predictive
+    value and/or auxiliary variable to inform the imputed value. The method
+    can apply forward, backward, construction or forward from construction
+    imputation. The type of imputation used will vary for each non-respondent
+    in each period depending on whether data is available in the predictive
+    period
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+         Original dataframe.
+    target : str
+        Column name of values to be imputed.
+    period : str
+        Column name containing datetime information.
+    reference : str
+        Column name of unique Identifier.
+    strata : str
+        Column name containing strata information (sic).
+    auxiliary : str
+        Column name containing auxiliary information (sic).
+
+    Returns
+    -------
+    pd.DataFrame
+        Original dataframe with imputed values in the target column, and with
+        intermediate columns which were used for the imputation method.
+    """
+
+    # Saving args to dict, so we can pass same attributes to multiple functions
+    # These arguments are used from the majority of functions
+    # Remove df to allow pipe chain
+
+    default_columns = locals()
+
+    del default_columns["df"]
+
+    if default_columns["filters"] is not None:
+
+        df = flag_rows_to_ignore(df, default_columns["filters"])
+
+    # TODO: Is datetime needed? If so here is a good place to convert to datetime
+    df["date"] = pd.to_datetime(df["date"], format="%Y%m")
+
+    # TODO: Pre calculated links
+
+    df = (
+        df.pipe(wrap_flag_matched_pairs, **default_columns)
+        .pipe(wrap_shift_by_strata_period, **default_columns)
+        .pipe(wrap_calculate_imputation_link, **default_columns)
+        .pipe(
+            create_impute_flags,
+            **default_columns,
+            predictive_auxiliary="f_predictive_auxiliary"
+        )
+        # TODO: How we gonna set defaults?
+        .fillna(
+            {"f_link_question": 1.0, "b_link_question": 1.0, "construction_link": 1.0}
+        )
+        .pipe(generate_imputation_marker)
+        .pipe(wrap_get_cumulative_links, **default_columns)
+        .pipe(
+            create_and_merge_imputation_values,
+            **default_columns,
+            imputation_class="group",
+            marker="imputation_marker",
+            combined_imputation="imputed_value",
+            cumulative_forward_link="cumulative_f_link_question",
+            cumulative_backward_link="cumulative_b_link_question",
+            construction_link="construction_link",
+            imputation_types=("c", "fir", "bir", "fic")
+        )
+    )
+
+    df = df.reset_index(drop=True)
```

Comment:
Making a note for future reference, why does the index need to be reset? Does one of the called functions modify index

Reply:
Yes that's why, we sort the data quite a few times , we need to chat about sorting and resetting index , in my opinion resetting index  should be done within the function is sorted.

Reply:
Added TODO for this